### PR TITLE
:seedling: Convert Dangerous Workflow check to probes

### PR DESCRIPTION
**Pull request by AdamKorcz**
#### What kind of change does this PR introduce?
feature

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the new behavior (if this is a feature change)?**

This PR converts the Dangerous Workflows check to probes.

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
NONE

#### Does this PR introduce a user-facing change?
NONE

```release-note

```

### CODE REVIEW

**Comment by AdamKorcz** [comment]
@laurentsimon  @spencerschrock PTAL.

File: checks/evaluation/dangerous_workflow.go [hunk 1]
```diff
@@ -15,59 +15,79 @@
...
-		dl.Warn(&checker.LogMessage{
-			Path:    e.File.Path,
-			Type:    e.File.Type,
-			Offset:  e.File.Offset,
-			Text:    text,
-			Snippet: e.File.Snippet,
-		})
+	if hasDWWithScriptInjection(findings) {
+		return checker.CreateMinScoreResult(name,
+			"dangerous workflow patterns detected")
```
Comment (spencerschrock) [inline checks/evaluation/dangerous_workflow.go:52]: can this just be combined with the one above?

```go
if hasDWWithUntrustedCheckout(findings) || hasDWWithScriptInjection(findings)
```

File: checks/evaluation/dangerous_workflow.go [hunk 2]
```diff
@@ -15,59 +15,79 @@
...
+		hasDangerousWorkflowUntrustedCheckout.Probe,
+	}
+
+	if !finding.UniqueProbesEqual(findings, expectedProbes) {
+		e := sce.WithMessage(sce.ErrScorecardInternal, "invalid probe results")
 		return checker.CreateRuntimeErrorResult(name, e)
 	}
 
-	if r.NumWorkflows == 0 {
+	if !hasWorkflows(findings) {
```
Comment (spencerschrock) [inline checks/evaluation/dangerous_workflow.go:41]: do we need a separate probe for if there are workflows if we have a notapplicable outcome now?

File: checks/evaluation/dangerous_workflow.go [hunk 3]
```diff
@@ -15,59 +15,79 @@
...
-		case checker.DangerousWorkflowUntrustedCheckout:
-			text = fmt.Sprintf("untrusted code checkout '%v'", e.File.Snippet)
-		case checker.DangerousWorkflowScriptInjection:
-			text = fmt.Sprintf("script injection with untrusted input '%v'", e.File.Snippet)
-		default:
-			err := sce.WithMessage(sce.ErrScorecardInternal, "invalid type")
-			return checker.CreateRuntimeErrorResult(name, err)
-		}
+	if hasDWWithUntrustedCheckout(findings) {
+		return checker.CreateMinScoreResult(name,
```
Comment (spencerschrock) [inline checks/evaluation/dangerous_workflow.go:46]: logging the offending findings here?

File: probes/hasAnyWorkflows/def.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+id: hasAnyWorkflows
+short: Checks whether a project has any workflows.
+motivation: >
+  This probe is only used to determine whether Scorecard can make conclusive scoring for Dangerous Workflow-related probes.
+implementation: >
+  The probe checks whether the project has any workflows and does not consider whether they are dangerous. The probe checks the `NumWorkflows` of the raw results.
+outcome:
+  - If a project does not have any workflows, one finding with OutcomeNegative (0) is returned.
+  - If a project has any workflows, one finding with OutcomeNegative (1) is returned.
```
Comment (spencerschrock) [inline probes/hasAnyWorkflows/def.yml:23]: copy paste errror of outcomenegative(1), should be positive?

But see the other comment first about if we still need this probe.

File: probes/hasDangerousWorkflowScriptInjection/impl.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	r := raw.DangerousWorkflowResults
+
+	for _, e := range r.Workflows {
+		if e.Type == checker.DangerousWorkflowScriptInjection {
+			return negativeOutcome()
+		}
+	}
+
+	return positiveOutcome()
```
Comment (spencerschrock) [inline probes/hasDangerousWorkflowScriptInjection/impl.go:45]: we lose the location / snippet information. Let's keep it so we can make it part of the other results?

File: probes/hasDangerousWorkflowUntrustedCheckout/impl.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+	r := raw.DangerousWorkflowResults
+
+	for _, e := range r.Workflows {
+		if e.Type == checker.DangerousWorkflowUntrustedCheckout {
+			return negativeOutcome()
+		}
+	}
+
+	return positiveOutcome()
```
Comment (spencerschrock) [inline probes/hasDangerousWorkflowUntrustedCheckout/impl.go:45]: same comment about losing location/snippet info

**Review by spencerschrock — COMMENTED** [review]
Was a bit of a skim to highlight the big comments.

File: probes/hasAnyWorkflows/def.yml (hunk 4 above)
Comment (AdamKorcz) [reply probes/hasAnyWorkflows/def.yml]: Removed probe in https://github.com/ossf/scorecard/pull/3521/commits/f1ad11a3e563eea643aa230ff8b9ec5db0489b5d

File: checks/evaluation/dangerous_workflow.go (hunk 2 above)
Comment (AdamKorcz) [reply checks/evaluation/dangerous_workflow.go:39]: Removed in https://github.com/ossf/scorecard/pull/3521/commits/f1ad11a3e563eea643aa230ff8b9ec5db0489b5d

File: probes/hasDangerousWorkflowUntrustedCheckout/impl.go (hunk 6 above)
Comment (AdamKorcz) [reply probes/hasDangerousWorkflowUntrustedCheckout/impl.go]: Fixed in https://github.com/ossf/scorecard/pull/3521/commits/e0b7800511d17cc8bdf28d78089b83b639172401

File: probes/hasDangerousWorkflowScriptInjection/impl.go (hunk 5 above)
Comment (AdamKorcz) [reply probes/hasDangerousWorkflowScriptInjection/impl.go]: Fixed in https://github.com/ossf/scorecard/pull/3521/commits/e0b7800511d17cc8bdf28d78089b83b639172401

File: checks/evaluation/dangerous_workflow.go (hunk 1 above)
Comment (AdamKorcz) [reply checks/evaluation/dangerous_workflow.go]: Fixed in https://github.com/ossf/scorecard/pull/3521/commits/713d267afcdb751dfff53cbe73bcb0566c7a6b0f

**Review by raghavkaul — APPROVED** [review]
Looks like comments were addressed, thanks

**Pull request merged**